<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="linkage">
<title>linkage • linkage</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="linkage">
<meta property="og:description" content="linkage">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">linkage</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/Install.html">Install</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-vignettes">Vignettes</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-vignettes">
    <h6 class="dropdown-header" data-toc-skip>R package</h6>
    <a class="dropdown-item" href="../articles/LinkageR.html">Genes and regulatory peaks to find transcription factors</a>
    <a class="dropdown-item" href="../articles/Motif.html">Monogenic MOTIF analysis</a>
    <a class="dropdown-item" href="../articles/Enrichment.html">Pathway Enrichment Module</a>
    <h6 class="dropdown-header" data-toc-skip>Shiny APP</h6>
    <a class="dropdown-item" href="../articles/Linkage.html">Linkage in shiny APP</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/aicplane/Linkage/issues">FAQ</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="http://www.xulabgdpu.org.cn/linkage">Shiny App</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://aicplane.github.io/Linkage/">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/aicplane/Linkage">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>linkage</h1>
            
      
      
      <div class="d-none name"><code>Linkage.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://aicplane.github.io/Linkage-tutorial/" class="external-link">linkage</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Linkage.UIO.html">Linkage.UIO</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="data-input">
<strong>Data input</strong><a class="anchor" aria-label="anchor" href="#data-input"></a>
</h3>
<div align="justify">
<p>The <strong><em>TCGA Breast Cancer cohort</em></strong> dataset
contains chromatin accessibility and gene expression data from 72
patients with breast cancer. The chromatin accessibility data is a
normalized bulk ATAC-seq count matrix, which a prior count of 5 is added
to the raw counts, then put into a “counts per million”, then log2
transformed, then quantile normalized. The gene expression data is a
normalized bulk RNA-seq count matrix, which is log2(fpkm+1)
transformed.</p>
</div>
<div align="justify">
<p>The <strong><em>GSE121589</em></strong> dataset contains chromatin
accessibility and gene expression data from 33 mice Muscle Stem Cells
(MuSCs). The chromatin accessibility data is a normalized bulk ATAC-seq
count matrix, which a prior count of 5 is added to the raw counts, then
put into a “counts per million”, then log2 transformed, then quantile
normalized. The gene expression data is a normalized bulk RNA-seq count
matrix, which is log2(fpkm+1) transformed.</p>
</div>
<div align="justify">
<p>Two tab-delimited text/csv input files (chromatin accessibility
matrix and gene expression matrix) are required before running
Linkage. The gene expression matrix file is a tab-delimited multi-column
data matrix, which the first column represents gene symbols and the
following columns represent normalized or raw expression levels of genes
for each sample. The chromatin accessibility matrix file is a
tab-delimited multi-column data matrix as well, which the first three
columns represent chromosome name, start coordinate on the chromosome
and end coordinate on the chromosome of the peaks respectively; the
remaining columns of the chromatin accessibility matrix file represent
normalized or raw chromatin accessibility levels of peaks for each
sample. <strong>Each element of ready-to-analysis files will appear in
Chromatin Accessibility Matrix panel and Gene Expression Matrix
panel.</strong></p>
</div>
<p><img src="../reference/figures/data.png"></p>
</div>
<div class="section level3">
<h3 id="regulatory-peaks-search-module">
<strong>Regulatory Peaks Search Module</strong><a class="anchor" aria-label="anchor" href="#regulatory-peaks-search-module"></a>
</h3>
<div align="justify">
<p><strong>Regulatory Peaks Search Module</strong> allows users to
detect all potential regulatory DNA regions for specific genes. When
given an <strong>input gene</strong> and <strong>targeting DNA
regions</strong>, Linkage will automatically perform <strong>canonical
correlation analysis</strong> between each quantitative chromatin
accessibility measure in the region and the quantitative expression
level of the gene across all samples. Users can easily adjust the region
of interest and customize the correlation algorithm
(<strong>Spearman</strong> / <strong>Pearson</strong> /
<strong>Kendall</strong>). Then, all the statistically significant
results are listed in the Potential Cis-regulatory Regions panel.
With clicking on a specific region of this panel, users can view the
scatter plot of quantitative chromatin accessibility and gene expression
from the Correlation Plot panel. The <strong>rho</strong> and
<strong>FDR</strong> for correlation analysis also be shown on the
scatter plot.</p>
</div>
<p><img src="../reference/figures/model_1.png"></p>
</div>
<div class="section level3">
<h3 id="regulatory-peaks-visualization-module">
<strong>Regulatory Peaks Visualization Module</strong><a class="anchor" aria-label="anchor" href="#regulatory-peaks-visualization-module"></a>
</h3>
<div align="justify">
<p><strong>Regulatory Peaks Visualization Module</strong> allows users
to visualize the coverage of mapped ATAC-seq reads around a given
specific regulatory peak, as well as the corresponding quantitative
expression of the target gene of this regulatory peak. Users initially
<strong>select a regulatory peak</strong> that obtained from the
Regulatory Peaks Detection Module. Linkage then categorizes samples into
five groups based on the quantitative chromatin accessibility of
the specific regulatory peak, ranging from low to high for each
individual sample. The <strong>coverage track of mapped ATAC-seq
reads</strong> and the <strong>boxplot of the target gene</strong> for
each group will be shown simultaneously.</p>
</div>
<p><img src="../reference/figures/model_2.png"></p>
</div>
<div class="section level3">
<h3 id="regulatory-peaks-annotation-module">
<strong>Regulatory Peaks Annotation Module</strong><a class="anchor" aria-label="anchor" href="#regulatory-peaks-annotation-module"></a>
</h3>
<div align="justify">
<p><strong>Regulatory Peaks Annotation Module</strong> allows users to
visualize the annotation of the predicational regulatory peaks from
Linkage. Once users click <strong>‘Annotate Regulatory Peaks’</strong>,
Linkage will perform the annotation of all predicational regulatory
peaks in terms of genomic location features, which includes whether the
peak is in the TSS, Exon, 5’ UTR, 3’ UTR, Intronic or Intergenic and
the position and strand information of the nearest gene of the peaks. To
effectively visualize the overlaps and distribution in annotation for
peaks, Linkage also produces the <strong>upsetplot</strong> that
adopting from the <strong>ChIPseeker</strong> package.</p>
</div>
<p><img src="../reference/figures/model_3.png"></p>
</div>
<div class="section level3">
<h3 id="motif-analysis-module">
<strong>Motif Analysis Module</strong><a class="anchor" aria-label="anchor" href="#motif-analysis-module"></a>
</h3>
<div align="justify">
<p><strong>Motif Analysis Module</strong> supports users to visualize
the enriched transcription factor motifs within potential regulatory
peaks. With clicking on a specific peak, users can view the location and
binding score information of each enriched transcription factor of this
region. Once users select one of the transcription factors, the
corresponding <strong>sequence logo</strong> of this cis-regulatory
element will appear in the Sequence-logo Plot panel.</p>
</div>
<p><img src="../reference/figures/model_4.png"></p>
</div>
<div class="section level3">
<h3 id="regulatory-network-module">
<strong>Regulatory Network Module</strong><a class="anchor" aria-label="anchor" href="#regulatory-network-module"></a>
</h3>
<div align="justify">
<p><strong>Regulatory Network Module</strong> helps users to visualize
GRNs that nodes are genes and corresponding cis-regulatory
elements，which inferred from previous analysis of Linkage. First,
users can input a list of interested genes or upload a gene-list file
that obtained from the previous analysis. Then users can adjust a series
of parameters that related to building the GRN, including types of gene
symbols, calculation methods and thresholds of interactions between the
nodes (edges of the GRNs). Once users click <strong>‘Build
Network’</strong>, Linkage will perform canonical correlation analysis
of quantitative expression level between each interested gene and
their potential cis-regulatory elements. The significant calculation
results of correlation analysis are shown in the Gene-TF Table
panel. Meanwhile, Linkage produces the corresponding informatic and
interactive GRN that adopting from the
<strong>visNetwork</strong> package. Users can further easily change
network layouts, select subnetworks, and save the GRN as
spreadsheets with interaction score or plots.</p>
</div>
<p><img src="../reference/figures/model_5_1.png"><img src="../reference/figures/model_5_2.png"></p>
</div>
<div class="section level3">
<h3 id="pathway-enrichment-module">
<strong>Pathway Enrichment Module</strong><a class="anchor" aria-label="anchor" href="#pathway-enrichment-module"></a>
</h3>
<div align="justify">
<p><strong>Pathway Enrichment Module</strong> supports users to
visualize tabular and graphical pathway enrichment results of interested
gene lists, which can be produced from previous analysis of Linkage.
The enrichment analysis can link these gene lists with underlying
molecular pathways and functional categories such as <strong>gene
ontology (GO)</strong> and <strong>Kyoto Encyclopedia of Genes and
Genomes (KEGG)</strong>.</p>
</div>
<p><img src="../reference/figures/model_6_1.png"><img src="../reference/figures/model_6_2.png"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Zenghui Liu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
